Don't import C++ class members that are protected or private

martinboehme opened the pull request:
Instead of mapping C++ access specifiers to Swift access levels, we simply skip importing private and protected members at all. They should never be used, so leaving them out removes clutter from the module interface.

We omit protected members in addition to private members because Swift structs can't inherit from C++ classes, so there's effectively no way to access them.

martinboehme committed:
Don't import C++ class members that are protected or private.

We omit protected members in addition to private members because Swift
structs can't inherit from C++ classes, so there's effectively no way to
access them.

CodaFi commented on tools/swift-ide-test/swift-ide-test.cpp:
```diff
@@ -691,6 +691,11 @@ static llvm::cl::opt<bool>
                        llvm::cl::desc("Disable ObjC interop."),
                        llvm::cl::cat(Category), llvm::cl::init(false));
 
+static llvm::cl::opt<bool>
```
Do you have IDE tests that require this flag?

martinboehme replied:
Yes -- see the test access-specifiers-module-interface.swift added in this PR.

Or do you mean a test that actually exercises features used in an IDE? In that case, no. My understanding, however, is that %target-swift-ide-test is the standard way of creating module interface tests?

CodaFi replied:
Ah, good. My eyes scanned over it - not used to seeing swift-ide-test being used to dump interfaces outside of the IDE test subdirectory...

CodaFi commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -112,6 +112,11 @@ static AccessLevel getOverridableAccessLevel(const DeclContext *dc) {
   return (dc->getSelfClassDecl() ? AccessLevel::Open : AccessLevel::Public);
 }
 
+static bool privateOrProtected(const clang::Decl *decl) {
```
We already have `importer::shouldSuppressDeclImport`. If you put this check there, you can nuke all of the extra stuff you've added here, and you'll handle the lazy importing cases for free.

martinboehme replied:
Thanks for the pointer -- doing this in a single place sounds appealing!

Unfortunately, when I tried this (i.e. returning true for private and protected members in `importer::shouldSuppressDeclImport` and removing the other code I had for this), it didn't work -- my tests started failing because the private members had become visible.

As far as I can tell, this is because `ClangImporter::Implementation::lookupValue` is only called for top-level declarations -- e.g. for the `PublicPrivate` class in my test. The class itself is, of course, imported. When my test then tries to access private members of `PublicPrivate`, I think Swift goes looking for them directly in the AST, which still contains these members. (I should say I'm new to the codebase, so my understanding may be wrong.)

I think this could be resolved by putting a call to `shouldSuppressDeclImport` in `ClangImporter::Implementation::importDeclImpl`. I'm not sure if this would have undesired consequences though -- maybe there's a reason this isn't being done yet? IOW, maybe we want to prevent the declarations from being added to the lookup table, but we still want them to be in the AST -- do you know?

martinboehme replied:
Having thought about this again, I've realized that while maybe I can't put `shouldSuppressDeclImport` in `ClangImporter::Implementation::importDeclImpl`, I could definitely do the access specifier check there -- see the commit I just pushed.

WDYT?

martinboehme committed:
Check access specifiers centrally in importDeclImpl().

CodaFi reviewed (approved):
I love it.  My one comment is that the ClangImporter bit could use a comment (I thought your old commit had a nice one, but it’s been a bit).

That shouldn’t block the merge.  LGTM.

CodaFi commented:
@swift-ci smoke test

martinboehme commented:
> I love it. My one comment is that the ClangImporter bit could use a comment (I thought your old commit had a nice one, but it’s been a bit).

I think you're probably just remembering my commit message -- but I agree that a comment would be useful.

I see the macOS build bot has failed anyway, so I'll add the comment (probably something along the lines of my commit message) along with the fix for the macOS build failure.

CodaFi commented:
Sounds good.

By the by, Windows just always fails the first trigger after force pushes since it doesn’t yet have the same auto-rescheduling behavior of the other bots. It’s also not a hard gate yet, so...

martinboehme committed:
Fix macOS build by using <stddef.h> instead of <cstddef>.

Apparently, the macOS toolchain doesn't provide <cstddef>.

<stddef.h> is used in test/Inputs/clang-importer-sdk/usr/include/macros.h,
so I'm assuming it will be OK. (I don't unfortunately have a macOS
machine to test on.)

martinboehme committed:
Add comment explaining why we skip private and protected C++ class
members.

martinboehme commented:
> Sounds good.
> 
> By the by, Windows just always fails the first trigger after force pushes since it doesn’t yet have the same auto-rescheduling behavior of the other bots. It’s also not a hard gate yet, so...

OK, thanks for the explanation!

gribozavr commented:
@swift-ci Please smoke test

compnerd commented:
This seems to have caused a regression on Windows.  Could you please take a look?  (CC @codafi @gribozavr)

compnerd commented:
@shahmishal could you please look into the triggering issue?  This actually would have caught the regression (the previous run definitely identified the issue)

shahmishal commented:
This was triggered and commit was updated - https://ci-external.swift.org/job/swift-PR-windows/826/
<img width="587" alt="Screen Shot 2020-03-19 at 5 41 28 PM" src="https://user-images.githubusercontent.com/2727770/77127206-e4417880-6a08-11ea-9ff1-2dfcee61dc09.png">

martinboehme commented:
> This seems to have caused a regression on Windows. Could you please take a look? (CC @CodaFi @gribozavr)

Sorry about the breakage. This is the relevant build bot, correct?

CodaFi commented:
Right. [These logs contain an example of the failure](https://ci-external.swift.org/job/swift-PR-windows/869/console).  Enums are default to being backed by signed on windows, so the check lines for the dump need to have generalized matches, the test needs to be special cased for Windows, or an explicit type bound should be provided to force the importer’s hand.

Up to you.

CodaFi commented:
I’ve got no idea why the headers are leaking C++ 17 code into the test, but that seems more pressing.

martinboehme commented:
> Right. [These logs contain an example of the failure](https://ci-external.swift.org/job/swift-PR-windows/869/console). Enums are default to being backed by signed on windows, so the check lines for the dump need to have generalized matches, the test needs to be special cased for Windows, or an explicit type bound should be provided to force the importer’s hand.
> 
> Up to you.

Thanks for explaining -- that had me scratching my head.

In the meantime, I see this PR is being reverted (and you've taken care of issues that occurred during the revert -- thanks!), so there's no rush here, correct?

I'm a bit confused by another of the failures (the one in confusables.swift) because it doesn't even use interop -- and I'm not changing anything outside of ClangImporter. Relevant snippet in the logs:

```
// expected-note @-1 {{required by referencing operator function '==' on 'BinaryInteger' where 'Self' = '(Int, Int)'}}

                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                       required by operator function '==' where 'Base' = '(Int, Int)'
```

Any ideas why that might have happened?

martinboehme commented:
> I’ve got no idea why the headers are leaking C++ 17 code into the test, but that seems more pressing.

Can you explain? I'm not sure exactly what you're referring to?

This isn't the issue related to `PathCch.h` though? That also seems like a fundamental issue to solve. What I think is happening is that this module is marked as "extern c" in the module map, so is being imported in C mode, but in `-enable-cxx-interop` mode, the `__cplusplus` is still being defined, so Clang is rightly complaining about conflicts between the const and non-const overloads of the functions. Does that sound plausible? I'll take a closer look at what's going on there.

martinboehme commented:
That said, I'm confused why my test would be pulling in `PathCch.h` in the first place... but it seems like a problem that needs fixing in any case.

martinboehme commented:
Having thought about the `PathCch.h` issue some more, I've realized my earlier comment didn't make a lot of sense.

`PathCch.h` is a header that contains its own `extern "C"` block (conditioned by an `#if __cplusplus`) for the C API that it exports and then provides additional C++-only APIs (again conditioned by an `#if __cplusplus`).

Because the entire `WinSDK` module is marked with the `[extern_c]` attribute, the whole of `PathCch.h` (which is contained in a submodule of the `WinSDK` module) is effectively wrapped in an `extern "C"` block. Of course, this doesn't work -- the C++-only APIs can't be placed in an `extern "C"` block because then they won't be name-mangled, and they'll conflict with any C APIs of the same name.

I think the right solution is to remove the `[extern_c]` attribute from the main `WinSDK` module and apply it only to those submodules that don't have `extern "C"` blocks in their headers already. I'll prepare a PR that does this.

martinboehme commented:
https://github.com/apple/swift/pull/30527 for the `PathCch.h` issue -- pretty sure this is the right fix.
